subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -295,33 +298,40 @@ private CompleteFileSystemLocationSnapshot snapshotFile(Path absoluteFilePath, S
         protected FileVisitResult doVisitFileFailed(Path file, IOException exc) {
-            // File loop exceptions are ignored. When we encounter a loop (via symbolic links), we continue
-            // so we include all the other files apart from the loop.
-            // This way, we include each file only once.
-            if (isNotFileSystemLoopException(exc)) {
-                String internedName = intern(file.getFileName().toString());
-                boolean isDirectory = Files.isDirectory(file);
-                if (shouldVisit(file, internedName, isDirectory, builder.getRelativePath())) {
-                    LOGGER.info("Could not read file path '{}'.", file);
-                    String internedAbsolutePath = intern(file.toString());
-                    builder.visitFile(new MissingFileSnapshot(internedAbsolutePath, internedName, AccessType.DIRECT));
+            relativePathTracker.enter(getInternedFileName(file));
+            try {
+                // File loop exceptions are ignored. When we encounter a loop (via symbolic links), we continue
+                // so we include all the other files apart from the loop.
+                // This way, we include each file only once.
+                if (isNotFileSystemLoopException(exc)) {
+                    String internedName = intern(file.getFileName().toString());
+                    boolean isDirectory = Files.isDirectory(file);
+                    if (shouldVisit(file, internedName, isDirectory, relativePathTracker.getSegments())) {
+                        LOGGER.info("Could not read file path '{}'.", file);
+                        String internedAbsolutePath = intern(file.toString());
+                        builder.visitLeafElement(new MissingFileSnapshot(internedAbsolutePath, internedName, AccessType.DIRECT));
+                    }
                 }
+                return FileVisitResult.CONTINUE;
+            } finally {
+                relativePathTracker.leave();
             }
-            return FileVisitResult.CONTINUE;
         }
 
         @Override
         protected FileVisitResult doPostVisitDirectory(Path dir, IOException exc) {
+            String dirName = relativePathTracker.leave();
             // File loop exceptions are ignored. When we encounter a loop (via symbolic links), we continue
             // so we include all the other files apart from the loop.
             // This way, we include each file only once.
             if (isNotFileSystemLoopException(exc)) {
                 throw new UncheckedIOException(String.format("Could not read directory path '%s'.", dir), exc);
             }
+            String absolutePath = intern(remapAbsolutePath(dir));
             AccessType accessType = AccessType.viaSymlink(
-                !symbolicLinkMappings.isEmpty() && symbolicLinkMappings.getFirst().target.equals(dir.toString())
+                !symbolicLinkMappings.isEmpty() && symbolicLinkMappings.getFirst().target.equals(absolutePath)
```

Comment:
IIUC, this needs to remain `dir.toString()`. The `absolutePath` is already the remapped path. Did no tests fail after this change?

Reply:
Why would this fail? This is just a string passed to `equals()`, it shouldn't matter if it's interned or not.

Reply:
Because it is a different string, isn't it?

Reply:
Ah, it's the remapping...

Reply:
Reverted.

Reply:
We should have a test which shows that this is broken. Can you add one?